xayn-ai/src/ltr/list_net/mod.rs

```diff
@@ -494,4 +970,301 @@ mod tests {
+    }
+
+    impl VecDataSource {
+        fn new(
+            training_data: Vec<(Array2<f32>, Array1<f32>)>,
+            evaluation_data: Vec<(Array2<f32>, Array1<f32>)>,
+        ) -> Self {
+            Self {
+                training_data_idx: 0,
+                training_data,
+                evaluation_data_idx: 0,
+                evaluation_data,
+            }
+        }
+    }
+
+    impl DataSource for VecDataSource {
+        fn reset(&mut self) {
+            self.training_data_idx = 0;
+            self.evaluation_data_idx = 0;
+        }
+
+        fn next_training_batch(&mut self, batch_size: usize) -> Vec<Sample> {
+            assert!(batch_size > 0);
+            let end_idx = self.training_data_idx + batch_size;
+            if end_idx <= self.training_data.len() {
+                let start_idx = self.training_data_idx;
+                self.training_data_idx = end_idx;
+
+                self.training_data[start_idx..end_idx]
+                    .iter()
+                    .map(|(inputs, target_prop_dist)| Sample {
+                        inputs: inputs.view(),
+                        target_prob_dist: target_prop_dist.view(),
+                    })
+                    .collect()
+            } else {
+                Vec::new()
+            }
+        }
+
+        fn next_evaluation_sample(&mut self) -> Option<Sample> {
+            if self.evaluation_data_idx < self.evaluation_data.len() {
+                let idx = self.evaluation_data_idx;
+                self.evaluation_data_idx += 1;
+
+                let data = &self.evaluation_data[idx];
+                Some(Sample {
+                    inputs: data.0.view(),
+                    target_prob_dist: data.1.view(),
+                })
+            } else {
+                None
+            }
+        }
+    }
+
+    struct TestCallbacks {
+        evaluation_results: Vec<Option<f32>>,
+    }
+
+    impl TestCallbacks {
+        fn new() -> Self {
+            Self {
+                evaluation_results: Vec::new(),
+            }
+        }
+    }
+
+    impl Callbacks for TestCallbacks {
+        fn begin_of_batch(&mut self) {
+            dbg!("begin batch");
+        }
+
+        fn end_of_batch(&mut self, losses: Vec<f32>) {
+            dbg!(losses);
+        }
+
+        fn begin_of_epoch(&mut self, _list_net: &ListNet) {
+            dbg!("begin epoch");
+        }
+
+        fn end_of_epoch(
+            &mut self,
+            _list_net: &ListNet,
+            mean_kv_divergence_evaluation: Option<f32>,
+        ) {
+            dbg!(mean_kv_divergence_evaluation);
+            self.evaluation_results.push(mean_kv_divergence_evaluation);
+        }
+
+        fn begin_of_training(&mut self, _list_net: &ListNet) {
+            dbg!("begin of training");
+        }
+
+        fn end_of_training(&mut self, _list_net: &ListNet) {
+            dbg!("end of training");
+        }
+    }
+
+    //FIXME create better tests
```

Comment:
it would be nice to have a set of tests which directly compares the result of a single forward and a backward pass for one input batch against the expected result.

Reply:
I have that in another branch and it's problematic.

The problem is that FFN learn to "cope with rounding errors" (oversimplified), but this means that over enough iterations even just the difference of rounding errors between different matmul implementations can make the network diverge to a point where they are comparable (but potentially still yield similar good results).

Even for a single batch with a single sample the results can diverge noticeable between the rust and the tensorflow implementations, reasons include some small hidden implementation differences around tensorflow clipping some values in some place and some other think.

So the test was "flaky"  dependent on the weights used for randomly initializing it, because of which it's currently not included as it at best would be misleading (looking like a test which show that it works but actually only working due to the initial weights chosen).